### PR TITLE
muon: init at unstable-2022-09-24

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
Current problems is:

- [] libpkgconf is not being found.

**Comment by AndersonTorres** [comment]
I believe the error is not from muon.
One week and I will merge this.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -16720,6 +16720,12 @@ with pkgs;
 
   samurai = callPackage ../development/tools/build-managers/samurai { };
 
+  muon = callPackage ../development/tools/build-managers/muon { };
+  muonStandalone = muon.override {
+    embedSamurai = true;
+    buildDocs = false;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:16726]: Why do we not need docs here?

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    ./stage-3/muon -C stage-3 test
+
+    runHook postCheck
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    stage-3/muon -C stage-3 install
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:124]: This is ignoring nix flags like -j$NIX_BUILD_CORES.

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  # Courtesy of sternenseemann and FRidh
+  featureFlag = feature: flag:
+    "-D${feature}=${if flag then "enabled" else "disabled"}";
+  conditionFlag = condition: flag:
+    "-D${condition}=${if flag then "true" else "false"}";
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "muon"
+          + lib.optionalString embedSamurai "-embedded-samurai"
+          + lib.optionalString static "-static";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:28]:
```suggestion
```
that doesn't really mater.

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+, samurai
+, scdoc
+, zlib
+, embedSamurai ? false
+, buildDocs ? true
+, static ? false
+}:
+
+let
+  # Courtesy of sternenseemann and FRidh
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:19]:
```suggestion
```

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  featureFlag = feature: flag:
+    "-D${feature}=${if flag then "enabled" else "disabled"}";
+  conditionFlag = condition: flag:
+    "-D${condition}=${if flag then "true" else "false"}";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:23]:
```suggestion
    "-D${condition}=${lib.boolToString flag}";
```

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+, libarchive
+, libpkgconf
+, pkgconf
+, python3
+, samurai
+, scdoc
+, zlib
+, embedSamurai ? false
+, buildDocs ? true
+, static ? false
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:15]:
```suggestion
```

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      ${featureFlag "samurai" embedSamurai} \
+      ${conditionFlag "static" static} \
+      stage-2
+    samu -C stage-2
+
+    stage-2/muon setup \
+      -Dprefix=$out \
+      ${featureFlag "docs" buildDocs} \
+      ${featureFlag "samurai" embedSamurai} \
+      ${conditionFlag "static" static} \
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:103]:
```suggestion
      ${conditionFlag "static" stdenv.isStatic} \
      stage-2
    samu -C stage-2

    stage-2/muon setup \
      -Dprefix=$out \
      ${featureFlag "docs" buildDocs} \
      ${featureFlag "samurai" embedSamurai} \
      ${conditionFlag "static" stdenv.isStatic} \
```

File: pkgs/development/tools/build-managers/muon/default.nix [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  '';
+
+  meta = with lib; {
+    homepage = "https://muon.build/";
+    description = "An implementation of Meson build system in C99";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ AndersonTorres ];
+    platforms = platforms.unix;
+    broken = stdenv.isDarwin  # typical `ar failure`
+      or static; # fails because of non-static libraries
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/muon/default.nix:136]:
```suggestion
      or stdenv.isStatic; # fails because of non-static libraries
```

**Review by SuperSandro2000 — COMMENTED** [review]
commit message should be ``muon: init at unstable-2022-08-24``

**Comment by AndersonTorres** [comment]
> commit message should be `muon: init at unstable-2022-08-24`

It was because I was waiting for the next release - sometimes they update the code in a daily basis, so I would "fill" the date later.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:16726]: The idea is to use it as a minimal standalone binary that requires only a C compiler (and some libs written in C).

The docs require Python with pyYAML libs.

File: pkgs/development/tools/build-managers/muon/default.nix (hunk 7 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/muon/default.nix:103]: attribute isStatic missing

File: pkgs/development/tools/build-managers/muon/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/muon/default.nix:117]: installation is just a copy, no compilation needed.

**Comment by AndersonTorres** [comment]
I believe this is good enough for now.

**Pull request merged**